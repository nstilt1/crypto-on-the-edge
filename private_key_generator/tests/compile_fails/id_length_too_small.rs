use private_key_generator_docs::{BinaryId, CryptoKeyGenerator, VersioningConfig, KeyGenerator, typenum::consts::{U19, U5}};
use hkdf::hmac::{Hmac, KeyInit};
use rand_chacha::ChaCha8Rng;
use rand::rngs::OsRng;
use sha2::Sha256;

type TestId = BinaryId<
    U19, // IdLength: too small. The other parameters would require this to be at least 20 bytes. Breakdown:
    U5,  // MacLength:                      5 bytes
    3,   // MAX_PREFIX_LEN:                 3 bytes
    32,  // VERSION_BITS:                           + 32 bits + constant 2 bits
    56,  // TIMESTAMP_BITS:                         + 56 bits
    8    // TIMESTAMP_PRECISION_REDUCTION:
>;  //                           20 bytes = 8 bytes + 11 bytes + round up 1 byte

type InvalidVersionLifetimeConfig = VersioningConfig<
    3, // EPOCH can be any time between 0 and now
    600, // VERSION LIFETIME must be at 600 seconds
    false, 
    4
>;

fn main() {
    type K = KeyGenerator<Hmac<Sha256>, InvalidVersionLifetimeConfig, ChaCha8Rng, Sha256>;

    let mut k = K::new(&[48u8; 32], b"ff", Hmac::<Sha256>::new_from_slice(&[42u8; 32]).unwrap(), &mut [3u8; 32]);
    let _id = k.generate_keyless_id::<TestId>(&[], &[], None, None, &mut OsRng);
}