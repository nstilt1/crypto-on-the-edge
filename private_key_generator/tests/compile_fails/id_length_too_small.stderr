error[E0080]: evaluation of `<private_key_generator_docs::BinaryId<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UTerm, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UTerm, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B1>, 3, 32, 56, 8> as private_key_generator_docs::EncodedId>::MAC_LENGTH` failed
 --> src/id.rs
  |
  | ...=> [/* The IdLength must be large enough to hold `MacLength bytes + MAX_PREFIX_LENGTH bytes + (2 + VERSION_BITS + TIMESTAMP_BITS) worth of bytes` */][ MacLength::USIZE ]
  |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ index out of bounds: the length is 0 but the index is 5

note: the above error was encountered while instantiating `fn private_key_generator_docs::KeyGenerator::<sha2::digest::core_api::CoreWrapper<hkdf::hmac::HmacCore<sha2::digest::core_api::CoreWrapper<sha2::digest::core_api::CtVariableCoreWrapper<sha2::Sha256VarCore, private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UTerm, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, sha2::OidSha256>>>>, private_key_generator_docs::VersioningConfig<3, 600, false, 4>, rand_chacha::ChaCha8Rng, sha2::digest::core_api::CoreWrapper<sha2::digest::core_api::CtVariableCoreWrapper<sha2::Sha256VarCore, private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UTerm, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, sha2::OidSha256>>>::fill_id_hmac::<private_key_generator_docs::BinaryId<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UTerm, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UInt<private_key_generator_docs::typenum::UTerm, private_key_generator_docs::typenum::B1>, private_key_generator_docs::typenum::B0>, private_key_generator_docs::typenum::B1>, 3, 32, 56, 8>>`
   --> src/key_generator.rs
    |
    |         self.fill_id_hmac(&mut id, id_type, associated_data);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
